airflow/jobs/scheduler_job.py

```diff
@@ -774,25 +775,30 @@ def _execute(self) -> None:
     @provide_session
     def _update_dag_run_state_for_paused_dags(self, session: Session = NEW_SESSION) -> None:
         try:
-            paused_dag_ids = DagModel.get_all_paused_dag_ids()
-            for dag_id in paused_dag_ids:
-                if dag_id in self._paused_dag_without_running_dagruns:
-                    continue
-
-                dag = SerializedDagModel.get_dag(dag_id)
-                if dag is None:
-                    continue
-                dag_runs = session.query(DagRun).filter(
-                    DagRun.dag_id == dag_id,
+            dag = None
+            paused_runs = (
+                session.query(DagRun)
+                .join(DagRun.dag_model)
+                .join(TaskInstance)
+                .filter(
+                    DagModel.is_paused == expression.true(),
                     DagRun.state == DagRunState.RUNNING,
                     DagRun.run_type != DagRunType.BACKFILL_JOB,
                 )
-                for dag_run in dag_runs:
-                    dag_run.dag = dag
-                    _, callback_to_run = dag_run.update_state(execute_callbacks=False)
-                    if callback_to_run:
-                        self._send_dag_callbacks_to_processor(dag, callback_to_run)
-                self._paused_dag_without_running_dagruns.add(dag_id)
+                .having(DagRun.last_scheduling_decision <= func.max(TaskInstance.updated_at))
+                .group_by(DagRun)
+                .order_by(DagRun.dag_id)
+            )
+            for dag_run in paused_runs:
+                if dag is None or dag.dag_id != dag_run.dag_id:
+                    dag = SerializedDagModel.get_dag(dag_run.dag_id)
```

Comment:
Rather than using SerializedDagModel directly we should use ` DagBag` as that has caching built in.
```suggestion
                    dag = self.dagbag.get_dag(dag_run.dag_id, session=session)
```

Reply:
Ah, that'll let me remove the funky None behavior too! Thanks.